### DISCUSSION TITLE
Minimal Constraints files not define

### DISCUSSION THREAD

**Discussion by sandeepmohan** [apache/airflow#22166] [General]
### Describe the issue with documentation

So for a base install of Airflow [no providers or extras] where use case requires only BashOperator, was looking to install Airflow from pip.

The standard `pip3 install apache-airflow==2.2.3` works but results in web server issues not rendering GUI as the flask version we have in our artifactory is higher. Web Server launches but GUI not accessible. 

If we try the constraints file `https://raw.githubusercontent.com/apache/airflow/constraints-2.2.3/constraints-3.7.txt` with the pip install, it tries to locate various libraries associated with provider packages many of which we have no use for no is on our Artifactory.

We are trying to install with no access to the internet. 

Can a constraints file be defined (or can I be pointed to one) that has the minimum libraries defined for a working installation sans any operator other than Bash, Dummy and Python? This seems to be the mission of `https://raw.githubusercontent.com/apache/airflow/constraints-2.2.3/constraints-3.7.txt` but the file has a large number of provider libraries that is "more than bare minimum or core"...

Thanks in advance. 

Sandeep

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by eladkal** [comment]
I think you are confusing constraints with requirements.
constraints doesn't install anything.

**Comment by sandeepmohan** [comment]
@eladkal  Understood but I can't find a minimal requirements file either. Constraints because I wanted the right versions of the libraries in place and referenced in install. I know req or constraints - the install will blow up if I am referencing azure or some of the apache providers. Would you know if there is a req file which would get me a working version of Airflow 2.2.3 with just Dummy, Bash and Python operators? I can then validate those libraries are available on my artifactory before running the install

**Comment by potiuk** [comment]
This is a discussion, not issue.

@sandeepmohan . There is no "requirements.txt" in  Airflow (and will never be). It is not needed, because we have setup.py where are all the dependencies are specified. You should follow `setup.py` semantics not `requirements.txt`  - those are mutually exclusive (and actually you can easily get requirements.txt by using our constraints - see below).

When you install airflow you install main airflow and specify which "optional" extras you want.

So you can install core apache airflow like that (and THIS is your minimal installation):

```
pip install apache-airflow==2.2.4 \
  --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-2.2.4/constraints-3.6.txt"
```

If you want to get `requirements.txt` from that you can create an empty virtualenv, install airflow the way described above and run  `pip freeze` and you will get your `requirements.txt` - exactly the same way as for any other package out there. This is standard `pip` and packages behaviour. 

Also you can add optional features you might need. For example you can add google provider and async capabilities of airflow like that:

```
pip install apache-airflow[google,async]==2.2.4 \
  --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-2.2.4/constraints-3.6.txt"
```

And again you can `pip freeze` if you want fixed set of requirements for this combination. This is where `setup.py` approach is superior vs. `requirements.txt` because you can generate such `requirements.txt` for any combination of:
* Python version
* Airflow
* Any set of extras you need.

Detailed list of of all "extras" you can use this way is here: https://airflow.apache.org/docs/apache-airflow/stable/extra-packages-ref.html

Without any extra, you only install core airflow and all the necessary dependencies. You DO NOT blow the installation in any way by installing unnecessary stuff you install everything that is necessary. And it is much better than "requiremetnts.txt" for example because the installed package sets might be different for different python version (and setup.py takes care about it as well as about the extras).

And to add to what Elad wrote - the constraints file is just list of "golden versions" of dependencies that we know work well with given version of airlfow. And this is the only "valid" way of installing airflow. It will pull the right version of dependencies. As @eladkal mentioned - you are NOT installing everything that is specified in the constraints - you are limiting the packages that you are installing to the version specified there.

**Reply by potiuk** [reply]
And adding after all the lecture above - yes there is such minimum list:

https://raw.githubusercontent.com/apache/airflow/constraints-2.2.4/constraints-no-providers-3.6.txt

**Reply by potiuk** [reply]
We build this and freeze automatically with every release - we never had any use of it, but I think this is as close as you can get. It is automatically prepared using what I described above.

**Reply by potiuk** [reply]
```
#
# This constraints file was automatically generated on 2022-02-18T16:34:15Z
# via "eager-upgrade" mechanism of PIP. For the "v2-2-test" branch of Airflow.
# This variant of constraints install just the 'bare' 'apache-airflow' package build from the HEAD of
# the branch, without installing any of the providers.
#
```

**Reply by potiuk** [reply]
Let me know if find it useful please.

**Reply by sandeepmohan** [reply]
This is so very awesome and exactly what I was looking for. A million thanks!!

**Reply by sandeepmohan** [reply]
And thanks too for the pip freeze guidance.

**Reply by potiuk** [reply]
You are welcome :)
